lib/rewrite/rewrite-rename.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2021 Balabit
+ * Copyright (c) 2021 Kokan <kokaipeter@gmail.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "rewrite-rename.h"
+
+typedef struct _LogRewriteRename LogRewriteRename;
+
+struct _LogRewriteRename
+{
+  LogRewrite super;
+  NVHandle source_handle;
+  NVHandle destination_handle;
+};
+
+void
+log_rewrite_rename_set_source(LogRewrite *s, const gchar *source)
+{
+  LogRewriteRename *self = (LogRewriteRename *) s;
+
+  self->source_handle = log_msg_get_value_handle(source);
+}
+
+void
+log_rewrite_rename_set_destination(LogRewrite *s, const gchar *destination)
+{
+  LogRewriteRename *self = (LogRewriteRename *) s;
+
+  self->destination_handle = log_msg_get_value_handle(destination);
+}
+
+static void
+log_rewrite_rename_process(LogRewrite *s, LogMessage **pmsg, const LogPathOptions *path_options)
+{
+  LogRewriteRename *self = (LogRewriteRename *) s;
+
+  if (self->source_handle == self->destination_handle)
+    return;
+
+  log_msg_make_writable(pmsg, path_options);
+
+  gssize value_len = 0;
+  const gchar *value = log_msg_get_value_if_set(*pmsg, self->source_handle, &value_len);
+  if (!value)
+    return;
+
+  log_msg_set_value(*pmsg, self->destination_handle, value, value_len);
+  log_msg_unset_value(*pmsg, self->source_handle);
+}
```

Comment:
can we perhaps handle the special case where we only need to change the NVTable entry, instead of copying the data?

Reply:
btw, this might be more complex than warranted for this rename() use-case, so as long as we delegate this functionality to a lower level function where this might be added later, this is completely optional.

Reply:
I would go here with a simple rename function on log message level, and as you said later this can be improved. Likely with introducing shift.